File: Deliverables/HEAVYROUTE/src/main/java/com/heavyroute/resources/dto/RoadEventDTO.java
package com.heavyroute.resources.dto;

import com.heavyroute.resources.enums.EventSeverity;
import com.heavyroute.resources.enums.RoadEventType;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

/**
 * Data Transfer Object (DTO) per la rappresentazione di un evento o segnalazione stradale.
 * <p>
 * Viene utilizzato per comunicare la presenza di ostacoli, pericoli o interruzioni
 * sulla rete viaria. Questi dati sono processati dal motore di navigazione per
 * garantire la sicurezza dei trasporti eccezionali.
 * </p>
 * * @author Heavy Route Team
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
public class RoadEventDTO {

    /**
     * Identificatore univoco della segnalazione.
     */
    private Long id;

    /**
     * Tipologia dell'evento stradale (es. INCIDENT, CONSTRUCTION, WEATHER).
     * Definisce la natura dell'impedimento riscontrato.
     */
    private RoadEventType type;

    /**
     * Livello di gravità dell'evento (LOW, MEDIUM, HIGH, CRITICAL).
     * Se la gravità è {@code CRITICAL}, l'evento è considerato bloccante e richiede
     * un ricalcolo immediato del percorso.
     */
    private EventSeverity severity;

    /**
     * Descrizione testuale opzionale con dettagli aggiuntivi sulla segnalazione.
     */
    private String description;

    /**
     * Latitudine della posizione GPS in cui si è verificato l'evento.
     * Deve rispettare il vincolo OCL: [-90.0, 90.0].
     */
    private Double latitude;

    /**
     * Longitudine della posizione GPS in cui si è verificato l'evento.
     * Deve rispettare il vincolo OCL: [-180.0, 180.0].
     */
    private Double longitude;

    /**
     * Data e ora di scadenza prevista per l'evento.
     * Oltre questa data, la segnalazione non viene più considerata attiva dal sistema.
     */
    private LocalDateTime expiresAt;

    /**
     * Indica se l'evento è attualmente attivo in base alla data di scadenza.
     * Campo calcolato utile per il filtraggio rapido lato frontend.
     */
    private boolean active;

    /**
     * Indica se l'evento blocca completamente il transito per la categoria del mezzo.
     * Derivato dalla severità {@code CRITICAL}.
     */
    private boolean blocking;
}

File: Deliverables/heavyroute_app/lib/features/planner/presentation/model/roadevent_dto.dart
/**
 * Data Transfer Object (DTO) per la creazione di una segnalazione stradale.
 * <p>
 * Include i dati geografici e temporali necessari per allertare il motore
 * di navigazione riguardo ostacoli o pericoli sulla rete viaria.
 * </p>
 * @author Roman
 */
class RoadEventCreateDTO {
  /** Tipologia dell'evento (es. OBSTACLE, ACCIDENT). */
  final String type;

  /** Livello di gravità (LOW, MEDIUM, CRITICAL). */
  final String severity;

  /** Descrizione testuale opzionale dell'evento. */
  final String? description;

  /** Latitudine GPS compresa nel range [-90.0, 90.0]. */
  final double latitude;

  /** Longitudine GPS compresa nel range [-180.0, 180.0]. */
  final double longitude;

  /** Data e ora di inizio validità della segnalazione. */
  final DateTime validFrom;

  /** Data e ora di fine validità (se nullo, l'evento è a tempo indeterminato). */
  final DateTime? validTo;

  RoadEventCreateDTO({
    required this.type,
    required this.severity,
    this.description,
    required this.latitude,
    required this.longitude,
    required this.validFrom,
    this.validTo,
  });

  /**
   * Serializza il DTO in JSON gestendo la conversione delle date in formato ISO8601.
   * @return Mappa pronta per l'invio tramite POST.
   */
  Map<String, dynamic> toJson() => {
    'type': type,
    'severity': severity,
    'description': description,
    'latitude': latitude,
    'longitude': longitude,
    'validFrom': validFrom.toIso8601String(),
    'validTo': validTo?.toIso8601String(),
  };

  /**
   * Factory method per la ricostruzione dell'oggetto da dati JSON.
   */
  factory RoadEventCreateDTO.fromJson(Map<String, dynamic> json) {
    return RoadEventCreateDTO(
      type: json['type'] as String,
      severity: json['severity'] as String,
      description: json['description'] as String?,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      validFrom: DateTime.parse(json['validFrom']),
      validTo: json['validTo'] != null ? DateTime.parse(json['validTo']) : null,
    );
  }
}